server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java

```diff
@@ -1261,6 +1318,413 @@ public void clusterStateProcessed(String source, ClusterState oldState, ClusterS
         );
     }
 
+    /**
+     * Set of pending snapshots deletions whose deletion is already triggered
+     */
+    private final Set<SnapshotId> ongoingSnapshotsDeletions = ConcurrentCollections.newConcurrentSet();
+
+    // only used in tests
+    public boolean hasOngoingSnapshotsDeletions(SnapshotId snapshotId) {
+        return ongoingSnapshotsDeletions.contains(snapshotId);
+    }
+
+    /**
+     * Set of pending snapshots deletions whose deletion is conflicting with on-going restores/clones/repository clean up or repository
+     * missing or read-only. Those sets are used to identify the cluster state updates to process in case they resolve some conflict.
+     */
+    private final Set<SnapshotId> pendingDeletionsWithConflictingRestores = ConcurrentCollections.newConcurrentSet();
+    private final Set<SnapshotId> pendingDeletionsWithConflictingCleanUps = ConcurrentCollections.newConcurrentSet();
+    private final Set<SnapshotId> pendingDeletionsWithConflictingClones = ConcurrentCollections.newConcurrentSet();
+    private final Set<SnapshotId> pendingDeletionsWithConflictingRepos = ConcurrentCollections.newConcurrentSet();
+
+    /**
+     * Find snapshots to delete in the the cluster state and triggers explicit snapshot delete requests. This method attempts to detect
+     * conflicting situations where triggering the snapshot deletion would likely fail due to a concurrent snapshot operation. In such
+     * cases the snapshot deletion is not triggered as it should be triggered by subsequent cluster state updates once the conflicting
+     * situation is resolved.
+     *
+     * The repository name and uuid information are extracted from the {@link SnapshotDeletionsPending} entries in order to find the
+     * repository to execute the snapshot delete request against. If the repo uuid was known at the time the snapshot was added to
+     * {@link SnapshotDeletionsPending} we try to find the corresponding repository, or a repository with a missing uuid but the same
+     * name. If the repo uuid was not known at the time the snapshot was added to {@link SnapshotDeletionsPending}, we try to find a
+     * repository with the same name.
+     *
+     * @param event the current {@link ClusterChangedEvent}
+     */
+    private void triggerSnapshotsPendingDeletions(final ClusterChangedEvent event) {
```

Comment:
Right, I should have added the unit tests in the first place.

I extracted the logic of deleting the `SnapshotDeletionsPending` into a dedicated `SnapshotDeletionsPendingExecutor` so that the logic is more contained and more easily testable using unit tests without having to create a complete `SnapshotService`. 

Unit tests have been added in `SnapshotDeletionsPendingExecutorTests` where we check that deletion are correctly triggered after cluster state updates and that conflict are indeed resolved.